mypy/checker.py

```diff
@@ -1076,7 +1076,8 @@ def check_assignment(self, lvalue: Node, rvalue: Node, infer_lvalue_type: bool =
                 elif (is_literal_none(rvalue) and
                         isinstance(lvalue, NameExpr) and
                         isinstance(lvalue.node, Var) and
-                        lvalue.node.is_initialized_in_class):
+                        lvalue.node.is_initialized_in_class and
+                        not experiments.STRICT_OPTIONAL):
```

Comment:
If we don't want this behavior, the check should just be deleted because `None`s are subtypes of all other types when not in strict Optional mode, so this check is unnecessary.

That said, I think what we really want is to update this check to be more permissive rather than doing AST replacement at the parser stage.